### DISCUSSION TITLE
Feature idea: Typesafe(er) raw queries

### DISCUSSION THREAD

**Discussion by Akxe** [prisma/prisma#10383] [Q&A] [answered]
Hello, I was wondering, if some of you would also find this interesting.

I would propose that Prisma could output original names along with Prisma aliases of tables and their columns.

It would allow for safer querying of data in tables
```ts
import { user, post } from '@prisma/client/tables';

function columnOfTable(column: PrismaColumn<any, any>): string {
  return `${column.parent.original}.${column.original}`;
}

function columnAsAlias(column: PrismaColumn<any, any>, alias = column.alias): string {
  return `${columnOfTable(column)} as ${alias}`;
}

prisma.$queryRaw`SELECT
  ${columnAsAlias(user.columns.name)},
  ${columnAsAlias(post.columns.name, 'postTitle')}
FROM ${user.original}
LEFT JOIN ${post.original} ON (${columnOfTable(user.userID)} = ${columnOfTable(post.authorID)})
WHERE ${columnOfTable(user.userID)} = 50`;
```

Benefits:
* If a column is removed the compilation fails and broken code never gets into production
* If a column is renamed the compilation resolve to a new name from Prisma generated code
* Possible higher-order functions:

Given 
```ts
interface PrismaAliased<T extends string> {
  original: string;
  /** Can be same as original */
  alias: T;
}

// Prisma generates static const code, that means that the interface bellow is only used for higher order functions
interface PrismaColumn<N extends string, T> extends PrismaAliased<N> {
  parent: PrismaTable;
  type: T; // A runtime equivalent of TS type ex.: '' | 0 | false | null
  defaultValue?: T; // It would be very useful to share the default values of DB and frontënd forms
}

// Prisma generates static const code, that means that the interface bellow is only used for higher order functions
interface PrismaTable<T extends string> extends PrismaAliased<T> {
  columns: Record<string, PrismaColumn>;
}
```

Here is one example that should be then possible:
```ts
function $saferQuery(...select: ...PrismaColumn[]): Promise<{ /* correctly inferred type */ }>;
```

**Comment by austincrim** [comment, marked as answer]
Hey @Akxe 👋,

We currently have an [open feature request](https://github.com/prisma/prisma/issues/2880) that is similar to this. Would you mind adding your +1 or use case in that thread if it is not already covered?

Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
